aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -196,53 +196,99 @@ Tensor pinverse(const Tensor& self, double rcond) {
   return at::linalg_pinv(self, rcond, /*hermitian=*/false);
 }
 
-Tensor& linalg_matrix_rank_out(Tensor& result, const Tensor& self, optional<double> tol, bool hermitian) {
-  checkSameDevice("linalg_matrix_rank", result, self);
+// Computes the rank of 'input' and saves the result in-place in 'result'
+// 'hermitian' controls whether SVD or eigendecomposition is used for computing the singular values
+// 'atol' and 'rtol' are the absolute and relative tolerances, respectively.
+// TODO: this function can be made public, see: https://github.com/pytorch/pytorch/issues/54151
+static Tensor& linalg_matrix_rank_out_helper(const Tensor& input, const Tensor& atol, const Tensor& rtol, bool hermitian, Tensor& result) {
+  checkSameDevice("torch.linalg.matrix_rank", result, input);
+  checkSameDevice("torch.linalg.matrix_rank", atol, input, "atol");
+  checkSameDevice("torch.linalg.matrix_rank", rtol, input, "rtol");
   ScalarType output_type = ScalarType::Long;
-  checkLinalgCompatibleDtype("linalg_matrix_rank", result.scalar_type(), output_type);
+  checkLinalgCompatibleDtype("torch.linalg.matrix_rank", result.scalar_type(), output_type);
 
   // Matrices or batch of matrices are allowed
-  TORCH_CHECK(self.dim() >= 2, "linalg_matrix_rank: Expected as input a matrix or a batch of matrices, but got a tensor of size: ", self.sizes());
+  TORCH_CHECK(input.dim() >= 2, "torch.linalg.matrix_rank: Expected as input a matrix or a batch of matrices, but got a tensor of size: ", input.sizes());
+
+  TORCH_CHECK(!at::isComplexType(atol.scalar_type()),
```

Comment:
Why not restrict atol and rtol to be the same dtype as input?

Reply:
We discussed this in `torch.linalg.pinv` PR. The decision there was that we shouldn't restrict the type of tolerances if it works as is.
> If directly atol/rtol tensor is passed, then all types should be valid that allow multiplication with max_singular_value tensor (that can be only of type float or double) and that allow comparison with a tensor of type float or double. So I guess any floating and integer types (so everything that is not complex) should be valid here for any allowed input?

See https://github.com/pytorch/pytorch/pull/48399#discussion_r544480613

`test_matrix_rank_tol` demonstrates that types from `torch.testing.all_types()` work for the tolerances tensor.

Reply:
Sounds good. It's nice to know we're consistent ;)